File: services/dbt_web/frontend/src/api/contracts.ts
export type TargetName = "staging" | "vault" | "marts";
export type ArtifactName = "manifest.json" | "catalog.json" | "run_results.json" | "graph.js";

export interface RunJobRequest {
  selectors: string[];
  vars: Record<string, string>;
  full_refresh: boolean;
  defer: boolean;
  fail_on_test_failure: boolean;
}

export interface RunJobResponse {
  run_id: string;
  status: string;
  submitted_at?: string;
  target: TargetName;
}

export interface RunStatusResponse {
  run_id: string;
  status: string;
  started_at?: string;
  finished_at?: string;
  duration_sec?: number;
  job_name?: string;
  artifacts: string[];
}

export interface RunLogsResponse {
  run_id: string;
  lines: string[];
  truncated: boolean;
  updated_at?: string;
}

export interface ArtifactEnvelope {
  run_id: string;
  name: ArtifactName;
  etag?: string;
  size?: number;
  content_type?: string;
  cached: boolean;
  content?: Record<string, unknown> | string;
}

export interface RunSummary {
  target: string;
  run_id?: string;
  generated_at?: string;
  dbt_version?: string;
  elapsed_time?: number;
  results_total: number;
  results_passed: number;
  results_failed: number;
  status: string;
}

export interface RunsListResponse {
  total: number;
  items: RunSummary[];
}

export interface ModelItem {
  unique_id: string;
  name: string;
  resource_type: string;
  schema?: string;
  package_name?: string;
  tags?: string[];
  depends_on?: string[];
}

export interface ModelSearchResponse {
  total: number;
  items: ModelItem[];
}

export interface LineageNode {
  id: string;
  name: string;
  resource_type: string;
  schema?: string;
  package_name?: string;
  tags?: string[];
}

export interface LineageEdge {
  source: string;
  target: string;
}

export interface LineageGraphResponse {
  nodes: LineageNode[];
  edges: LineageEdge[];
  total_nodes: number;
  total_edges: number;
}

export interface TestFailure {
  target: string;
  unique_id: string;
  status: string;
  severity: string;
  message?: string;
  execution_time?: number;
}

export interface TestsSummaryResponse {
  summary: Record<string, { total: number; passed: number; failed: number }>;
  failed: TestFailure[];
  failed_total: number;
}

File: services/dbt_web/backend/app/schemas/contracts.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

ArtifactName = Literal["manifest.json", "catalog.json", "run_results.json", "graph.js"]
TargetName = Literal["staging", "vault", "marts"]


class RunJobRequest(BaseModel):
    selectors: list[str] = Field(default_factory=list)
    vars: dict[str, str] = Field(default_factory=dict)
    full_refresh: bool = False
    defer: bool = False
    fail_on_test_failure: bool = True


class RunJobResponse(BaseModel):
    run_id: str
    status: str
    submitted_at: datetime | None = None
    target: TargetName


class RunStatusResponse(BaseModel):
    run_id: str
    status: str
    started_at: datetime | None = None
    finished_at: datetime | None = None
    duration_sec: float | None = None
    job_name: str | None = None
    artifacts: list[str] = Field(default_factory=list)


class RunLogsResponse(BaseModel):
    run_id: str
    lines: list[str] = Field(default_factory=list)
    truncated: bool = False
    updated_at: datetime | None = None


class ArtifactEnvelope(BaseModel):
    run_id: str
    name: ArtifactName
    etag: str | None = None
    size: int | None = None
    content_type: str | None = None
    cached: bool = False
    content: dict | str | None = None


class EventPayload(BaseModel):
    event_id: str
    dag_id: str
    run_id: str | None = None
    event_ts: datetime
    target_layer: str | None = None
    upstream_run_ref: str | None = None


class ModelSearchResponse(BaseModel):
    total: int
    items: list[dict]
